src/power.jl

```diff
@@ -184,31 +211,33 @@ c_k & = & 2 \\sum_{j=0}^{(k-2)/2} a_{k-j} a_j + (a_{k/2})^2,
 """ sqr!
 
 for T = (:Taylor1, :TaylorN)
-    @eval function sqr!(c::$T, a::$T, k::Int)
-        if k == 0
-            @inbounds c[1] = a[1]^2
-            return nothing
-        end
+    @eval begin
+        @inline function sqr!(c::$T, a::$T, k::Int)
+            if k == 0
+                @inbounds c[1] = constant_term(a)^2
+                return nothing
+            end
+
+            kodd = k%2
+            kend = div(k - 2 + kodd, 2)
+            @inbounds for i = 0:kend
+                if $T == Taylor1
+                    c[k+1] += a[i+1] * a[k-i+1]
+                else
+                    mul!(c[k+1], a[i+1], a[k-i+1])
+                end
+            end
+            @inbounds c[k+1] = 2 * c[k+1]
+            kodd == 1 && return nothing
 
-        kodd = k%2
-        kend = div(k - 2 + kodd, 2)
-        @inbounds for i = 0:kend
             if $T == Taylor1
-                c[k+1] += a[i+1] * a[k-i+1]
+                @inbounds c[k+1] += a[div(k,2)+1]^2
```

Comment:
`sqr!` for scalars will be non-mutating, so you will have to return the updated value. Then, either you keep the `if`, or return a corresponding value for `sqr!` when it acts on `HomogenousPolynomials`; the latter will have a cost in memory...